src/iface/neighbor.rs

```diff
@@ -60,23 +61,59 @@ impl<'a> Cache<'a> {
     /// Neighbor entry lifetime, in milliseconds.
     pub(crate) const ENTRY_LIFETIME: Duration = Duration { millis: 60_000 };
 
+    /// Default number of entries in the cache before GC kicks in
+    pub(crate) const GC_THRESH: usize = 1024;
+
     /// Create a cache. The backing storage is cleared upon creation.
     ///
     /// # Panics
     /// This function panics if `storage.len() == 0`.
     pub fn new<T>(storage: T) -> Cache<'a>
             where T: Into<ManagedMap<'a, IpAddress, Neighbor>> {
+
+        Cache::new_with_limit(storage, Cache::GC_THRESH)
+    }
+
+    pub fn new_with_limit<T>(storage: T, gc_thresh: usize) -> Cache<'a>
+            where T: Into<ManagedMap<'a, IpAddress, Neighbor>> {
         let mut storage = storage.into();
         storage.clear();
 
-        Cache { storage, silent_until: Instant::from_millis(0) }
+        Cache { storage, gc_thresh: gc_thresh, silent_until: Instant::from_millis(0) }
+    }
+
+    #[cfg(any(feature = "std", feature = "alloc"))]
+    pub fn run_owned_gc(&mut self, timestamp: Instant) {
+        let new_map = match self.storage {
+            ManagedMap::Borrowed(_) => {
+                unreachable!()
+            }
+            ManagedMap::Owned(ref mut map) => {
+                map.iter_mut()
+                .map(|(key, value)| (*key, value.clone()))
+                .filter(|(_, v)| timestamp < v.expires_at)
+                .collect()
+            }
+        };
+        self.storage = ManagedMap::Owned(new_map);
     }
 
     pub(crate) fn fill(&mut self, protocol_addr: IpAddress, hardware_addr: EthernetAddress,
                        timestamp: Instant) {
         debug_assert!(protocol_addr.is_unicast());
         debug_assert!(hardware_addr.is_unicast());
 
+        match self.storage {
+            ManagedMap::Borrowed(_) =>  {
+                ();
+            }
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            ManagedMap::Owned(_) => {
+                if self.storage.len() >= self.gc_thresh {
+                    self.run_owned_gc(timestamp);
```

Comment:
Would it be more readable to have a static method that runs GC on a `BTreeMap` instead of `self`? That way we don't have the `match` with the `unreachable` statement. Just a thought.

Reply:
:+1: